tools/buildsys/src/gomod.rs

```diff
@@ -0,0 +1,176 @@
+include only the source code of the project, but not the source code of any
+dependencies. The Go programming language promotes the use of "modules" for
+dependencies and projects adopting modules will provide go.mod and go.sum
+files.
+
+This module extends the functionality of `packages.metadata.build-package.external-files`
+and provides the ability to retrieve and validate dependencies
+declared using Go modules given a tar archive containing a go.mod and go.sum.
+
+The location where dependencies are retrieved from are controlled by the
+standard environment variables employed by the Go tool: GOPROXY, GOSUMDB, and
+GOPRIVATE.
+
+ */
+
+pub(crate) mod error;
+use error::Result;
+
+use super::manifest;
+use duct::cmd;
+use snafu::{ensure, OptionExt, ResultExt};
+use std::env;
+use std::path::{Path, PathBuf};
+use std::process::Output;
+
+pub(crate) struct GoMod;
+
+impl GoMod {
+    pub(crate) fn vendor(
+        root_dir: &Path,
+        package_dir: &Path,
+        gomod: &manifest::ExternalFile,
+    ) -> Result<()> {
+        let url_file_name = extract_file_name(&gomod.url)?;
+        let path_arg = &gomod.path.as_ref().unwrap_or(&url_file_name);
+        ensure!(path_arg.components().count() == 1, error::InputFileSnafu);
+
+        let path = package_dir.join(path_arg);
+        ensure!(path.is_file(), error::InputFileBadSnafu { path });
+        let mod_dir = gomod.bundle_path.as_ref().context(error::ModDirSnafu)?;
+        let output_path_arg = gomod
+            .bundle_output_path
+            .as_ref()
+            .context(error::OutputDirSnafu)?;
+        let output_path = package_dir.join(output_path_arg);
+
+        // Return early if the output path exists and is a file, assuming it's already been built
+        if output_path.exists() && output_path.is_file() {
+            return Ok(());
+        }
+
+        // Our SDK and toolchain are picked by the external `cargo make` invocation.
+        let sdk = getenv("BUILDSYS_SDK_IMAGE")?;
+
+        // Several Go variables control proxying
+        let mut goproxy = go_env("GOPROXY").unwrap_or_else(|| "".to_string());
+        if goproxy.ends_with('\n') {
+            goproxy.pop();
+        }
+        let mut gosumdb = go_env("GOSUMDB").unwrap_or_else(|| "".to_string());
+        if gosumdb.ends_with('\n') {
+            gosumdb.pop();
+        }
+        let mut goprivate = go_env("GOPRIVATE").unwrap_or_else(|| "".to_string());
+        if goprivate.ends_with('\n') {
+            goprivate.pop();
+        }
+
+        let args = DockerGoArgs {
+            module_path: package_dir,
+            sdk_image: sdk,
+            go_mod_cache: &root_dir.join(".gomodcache"),
+            command: format!(
+                "tar xf {input} &&
+                pushd {moddir} &&
+                export GOPROXY={goproxy} &&
+                export GOSUMDB={gosumdb} &&
+                export GOPRIVATE={goprivate} &&
+                go list -mod=readonly ./... >/dev/null && go mod vendor &&
```

Comment:
This `go mod vendor` seems to be causing problems for me. For projects that just use `go.mod`, this causes a `vendor` directory to be created and everything to be downloaded (again) into that new directory. That seems to be the cause of these errors:

```
  #19 0.532 go: inconsistent vendoring in /home/builder/rpmbuild/BUILD/cloud-provider-aws-1.22.3/GOPATH/src/github.com/kubernetes/cloud-provider-aws:
  #19 0.532     github.com/aws/aws-sdk-go@v1.43.32: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
  #19 0.532     github.com/golang/mock@v1.4.4: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
  #19 0.532     github.com/google/go-cmp@v0.5.5: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
  #19 0.532     github.com/stretchr/testify@v1.7.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
...
```

Can we add a check somewhere that first checks if there is a `vendor` directory, and based on that decide whether to run `go mod vendor` or not?

Reply:
Maybe this isn't the actual source of the failure. I tried removing this locally, but I still get the error. So my error may be coming from elsewhere, or it may be left over cached files in my environment. If it's something left behind, I haven't been able to figure out yet where that would be or how to clean it up though.

But somewhere a `go mod vendor` is happening and seems to be the source of my problems.

Reply:
> Can we add a check somewhere that first checks if there is a vendor directory, and based on that decide whether to run go mod vendor or not?

If there is already an existing `vendor` directory, we shouldn't be calling out to `go mod` again. This would indicate that the upstream project has declared their dependencies, has them part of their source control, and the project is ready to be build via `go build`. I _believe_ this is already the current paradigm as evident by packages like `runc`: 

https://github.com/bottlerocket-os/bottlerocket/blob/211e3db67a7c0ac758c6eb135213285513c92f82/packages/runc/runc.spec#L34-L38

and their vendoring inside the project

https://github.com/opencontainers/runc/tree/main/vendor

Reply:
> If there is already an existing vendor directory, we shouldn't be calling out to go mod again

No, my point was if there ISN'T an existing `vendor` directory, then don't try to run `go mod vendor` as that's not the type of dependency management that project uses. But based on our offline conversation, sounds like this may still be wanted to be a way to "cache" those dependencies locally to allow things to be built in a network isolated container.

So maybe this is just a red herring and there's some other issue with the package I'm working on. Still trying to sort that out.

Reply:
> may still be wanted to be a way to "cache" those dependencies locally to allow things to be built in a network isolated container.

Right, this will enable packages being build in our network isolated `buildsys` container to not have to gather the dependencies manually (like hotdog _was_ doing) but rather, build a "bundled" tarball with the vendored dependencies to then be consumed in the isolated environment